replace `hint_common_parent_position()` by backwards accumulator updates

ces42 opened the pull request:
Only calls to `evaluate()` now trigger NNUE accumulator updates. To make sure that we are likely to find parent positions from which to update the accumulators we perform a backwards NNUE update whenever we compute the accumulator from scratch for some position.

passed STC
LLR: 2.93 (-2.94,2.94) <0.00,2.00>
Total: 39680 W: 10474 L: 10164 D: 19042
Ptnml(0-2): 171, 4068, 11042, 4398, 161 
https://tests.stockfishchess.org/tests/view/67a27f26eb183d11c65945be

passed LTC
LLR: 2.94 (-2.94,2.94) <0.50,2.50>
Total: 337308 W: 86408 L: 85550 D: 165350
Ptnml(0-2): 276, 30551, 106126, 31441, 260
https://tests.stockfishchess.org/tests/view/67a287efeb183d11c65945ee

then simplified:
STC
LLR: 2.94 (-2.94,2.94) <-1.75,0.25>
Total: 28608 W: 7641 L: 7413 D: 13554
Ptnml(0-2): 132, 3036, 7744, 3256, 136 
https://tests.stockfishchess.org/tests/view/67a4703719f522d3866d3345

LTC
LLR: 2.96 (-2.94,2.94) <-1.75,0.25>
Total: 200226 W: 51026 L: 50990 D: 98210
Ptnml(0-2): 170, 18468, 62799, 18508, 168 
https://tests.stockfishchess.org/tests/view/67a4f255229c1a170cc08964

The version in this PR is a bit different from the simplified version, but it's compile-time changes only.

xu-shawn commented:
Congratulations! Just FYI, an LTC is not strictly required for non-functional speedups.

ces42 commented:
well vondele told me to run LTC on discord, because of the size of the change. It's also not really a non-functional change since it changes the algorithm for computing evaluations, so it is conceivable that this would be slower at higher depth.

xu-shawn commented on src/nnue/nnue_feature_transformer.h:
```diff
@@ -468,38 +467,25 @@ class FeatureTransformer {
+                                        StateInfo*       target_state,
+                                        const StateInfo* computed) const {
+        constexpr bool Forward   = Direction == FORWARD;
+        constexpr bool Backwards = Direction == BACKWARDS;
```
maybe add a `[[maybe_unused]]` attribute to this?

xu-shawn commented:
Thread sanitizer seems to have failed. Maybe this has something to do with the shared root state?

ces42 commented:
Yeah, I'm looking at it.

mstembera commented:
Congrats!  Can you say a bit about why the small accumulator can't be handled the same way as the big one?

mstembera commented on src/position.h:
```diff
@@ -47,8 +48,8 @@ struct StateInfo {
-    int    rule50;
-    int    pliesFromNull;
+    int16_t    rule50;
+    int16_t    pliesFromNull;
```
Why are these 16 bit now?

vondele commented:
can you please squash the PR, also we'll need to understand why CI is / was failing, in particular if thread sanitizer is flagging something.

ces42 commented:
> Thread sanitizer seems to have failed. Maybe this has something to do with the shared root state?

Yes, that seems to be the issue. Or at least nodes where `st->previous->next != st` when doing SMP, which I assume are the children of the root node? Do you know of a good way to detect these? Adding the condition `st->previous->next == st` to the loop looking for a computed parent fixed the issue. But I am slightly worried about race conditions where the condition is OK when the parent node is searched for, but then the `next`-pointer of the parent is changed by another thread before the NNUE update is performed. Do you know of a better way to detect root nodes?

Incidentally, why do some root nodes not have `st->previous == nullptr`?

ces42 commented:
> Congrats! Can you say a bit about why the small accumulator can't be handled the same way as the big one?

Since there are much fewer smallnet evaluations (only 20% or so I think) the heuristic "computing the accumulator of an interior node has zero effective cost because we will be able to reuse it often" does not work for the small accumulator (The old code that does hints for the small accumulator is also definitively "wrong" in this sense). But the small accumulator is also 24x cheaper to compute so there's basically nothing to optimize there.

ces42 commented:
OK I have fixed the SMP issue. Sorry for the mess. This should be ready for merging now. I ran an SMP test on [8074a0b](https://github.com/official-stockfish/Stockfish/pull/5870/commits/8074a0b0b49f3a9f2b620f4dd14db3e1e32f2172)

SMP STC
LLR: 2.95 (-2.94,2.94) <0.00,2.00>
Total: 45464 W: 11977 L: 11662 D: 21825
Ptnml(0-2): 57, 5054, 12196, 5367, 58 
https://tests.stockfishchess.org/tests/view/67a7d1d00b25c820efc8bc2d

I have added two more commits since that test, but the first one (reverting [5aedb70](https://github.com/official-stockfish/Stockfish/pull/5870/commits/5aedb70f90f66c6e3be980cd53c798a677bf445b)) should be a strict speedup if anything and the second one is just changing the type of two ints and should be elo neutral (cf [this test](https://tests.stockfishchess.org/tests/view/67a3e72519f522d3866d3239)). 

should I run any tests on the final version?

I'm not squashing commits yet for reference purposes.

Sopel97 commented on src/nnue/nnue_feature_transformer.h:
```diff
@@ -539,12 +529,13 @@ class FeatureTransformer {
 
-            if (removed.size() == 1)
+            if ((Forward && removed.size() == 1)
+                || (Backwards && added.size() == 1))  // added.size() == removed.size() == 1
```
I'd move the code from these comments to asserts inside the ifs

Sopel97 commented on src/nnue/nnue_feature_transformer.h:
```diff
@@ -815,16 +828,41 @@ class FeatureTransformer {
+        {
+            if (FeatureSet::requires_refresh(st, Perspective)
+                || (!Big && (gain -= FeatureSet::update_cost(st) < 0)) || !st->previous || st->previous->next != st)
+                goto refresh;
```
I really don't like this goto, as removing it would only involve a boolean flag and `break`.

mstembera commented:
This https://tests.stockfishchess.org/tests/view/67abfe1ca04df5eb8dbebf0b passed.  I can open a PR after this gets merged but whatever is easiest is fine by me.

Disservin commented:
@ces42 will you resolve the `goto` comment as well?

Disservin commented:
fyi needs a rebase
